### PR TITLE
Added support for actions that run before&after snapshot roots and before/after specific folders

### LINKED ISSUE #400: FR: Add configurable before-upload/after-upload scripts for directories.
Use cases:

- when backing up certain directories, we may need to instruct some daemon to flush writes and hold them in memory until directory has been saved

- another use case would be taking ZFS or BTRFS point-in-time snapshots of a directory and working off that, instead of source data that keeps changing

- custom notifications after snapshot success/failure

This could be done via policies attached to FS nodes.

### PR DESCRIPTION
**Pull request by jkowalski**
Fixes #400

Related: #394 #709 #683

### CODE REVIEW

**Review by julio-lopez — COMMENTED** [review]
👍 LG +1

A couple of minor inline comments.

Also, a high-level question about the names / terminology?

Do we want to use the term _"hooks"_ or something else? Not saying it is wrong or suggesting something else. Merely posing the question because once this is merged, it becomes part of the API in the forms policy setting names and CLI flags, and it becomes harder to change down the road without breaking things for existing users.

File: cli/command_policy_set_hooks.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+	*changeCount++
+
+	if *policySetPersistHookScript {
+		script, err := ioutil.ReadFile(value) //nolint:gosec
+		if err != nil {
+			return err
+		}
+
+		if len(script) > maxScriptLength {
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply cli/command_policy_set_hooks.go:76]: probably not worth it, unless the concern is OOMing, which given this is CLI is not a big one.

**Review by julio-lopez — APPROVED** [review]

File: tests/end_to_end_test/snapshot_actions_test.go [hunk 2]
```diff
@@ -219,6 +220,50 @@ func TestSnapshotActionsBeforeAfterFolder(t *testing.T) {
...
+	t.Parallel()
+
+	e := testenv.NewCLITest(t)
+
+	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir)
+	defer e.RunAndExpectSuccess(t, "repo", "disconnect")
+
+	var (
+		successScript      = tmpfileWithContents(t, "echo Hello world!")
+		successScript2     = tmpfileWithContents(t, "echo Hello world!")
```
Comment (julio-lopez) [inline tests/end_to_end_test/snapshot_actions_test.go:233]: nit: for clarity, use the same pattern that is used for `failingScript` with `successScript2`.

That is, declare it a `string` here and set it in the `"windows"` block below.

Comment (jkowalski) [reply tests/end_to_end_test/snapshot_actions_test.go]: done

**Review by julio-lopez — APPROVED** [review]

**Pull request merged**